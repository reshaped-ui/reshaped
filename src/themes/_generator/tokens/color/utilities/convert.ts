import { useMode, modeOklch, type Oklch, formatHex, modeRgb } from "culori/fn";
import type * as TColor from "../color.types";

// eslint-disable-next-line react-hooks/rules-of-hooks
const oklch = useMode(modeOklch);
// eslint-disable-next-line react-hooks/rules-of-hooks
const rgb = useMode(modeRgb);

export const hexToOklch = (hex: string): Oklch => {
	const result = oklch(formatHex(hex));

	if (!result) throw new Error(`[Reshaped] Can't convert ${hex} to oklch`);
	return result;
};

export const oklchToRgb = (oklch: Oklch) => {
	const value = rgb(oklch);

	return {
		...value,
		r: Number(value.r.toFixed(4)),
		g: Number(value.g.toFixed(4)),
		b: Number(value.b.toFixed(4)),
	};
};

export const tokenToOklchToken = (token: TColor.Token): TColor.InternalToken => {
	const hexDarkFallback = token.hexDark ? hexToOklch(token.hexDark) : undefined;

	return {
		oklch: token.oklch ? { ...token.oklch, mode: "oklch" } : hexToOklch(token.hex),
		oklchDark: token.oklchDark ? { ...token.oklchDark, mode: "oklch" } : hexDarkFallback,
	};
};
