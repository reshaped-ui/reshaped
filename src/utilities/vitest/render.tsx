import { render as vitestRender } from "vitest-browser-react";
import Reshaped from "components/Reshaped";
import "themes/reshaped/theme.css";

const Wrapper = (props: { children: React.ReactNode }) => {
	return <Reshaped>{props.children}</Reshaped>;
};

const render = (node: React.ReactNode) => {
	const screen = vitestRender(<div id="root">{node}</div>, {
		wrapper: Wrapper,
	});

	return { ...screen, container: document.getElementById("root")! };
};

export default render;
