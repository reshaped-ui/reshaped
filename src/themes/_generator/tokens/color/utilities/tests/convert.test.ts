import { expect, test, describe } from "vitest";
import { hexToOklch, oklchToRgb, tokenToOklchToken } from "../convert";

const white = hexToOklch("#ffffff");
const black = hexToOklch("#000000");
const red = hexToOklch("#ff0000");

describe("themes/tokens/color/utilities/convert", () => {
	describe("hexToOklch", () => {
		test("white", () => {
			expect(white).toMatchSnapshot();
		});

		test("black", () => {
			expect(black).toMatchSnapshot();
		});

		test("red", () => {
			expect(red).toMatchSnapshot();
		});

		test("white, short", () => {
			expect(hexToOklch("#fff")).toEqual(white);
		});
	});

	describe("oklchToRgb", () => {
		test("white", () => {
			expect(oklchToRgb(white)).toMatchSnapshot();
		});

		test("black", () => {
			expect(oklchToRgb(black)).toMatchSnapshot();
		});

		test("red", () => {
			expect(oklchToRgb(red)).toMatchSnapshot();
		});
	});

	describe("tokenToOklchToken", () => {
		test("hex", () => {
			expect(tokenToOklchToken({ hex: "#ffffff" })).toMatchSnapshot();
		});

		test("hex and hexDark", () => {
			expect(tokenToOklchToken({ hex: "#ffffff", hexDark: "#000000" })).toMatchSnapshot();
		});

		test("oklch", () => {
			expect(tokenToOklchToken({ oklch: white })).toMatchSnapshot();
		});

		test("oklch and oklchDark", () => {
			expect(tokenToOklchToken({ oklch: white, oklchDark: black })).toMatchSnapshot();
		});
	});
});
