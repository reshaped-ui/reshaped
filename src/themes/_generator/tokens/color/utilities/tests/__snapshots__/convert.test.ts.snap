// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`themes/tokens/color/utilities/convert > hexToOklch > black 1`] = `
{
  "c": 0,
  "l": 0,
  "mode": "oklch",
}
`;

exports[`themes/tokens/color/utilities/convert > hexToOklch > red 1`] = `
{
  "c": 0.2576833038053608,
  "h": 29.233880279627854,
  "l": 0.6279553639214311,
  "mode": "oklch",
}
`;

exports[`themes/tokens/color/utilities/convert > hexToOklch > white 1`] = `
{
  "c": 0,
  "l": 1.0000000000000002,
  "mode": "oklch",
}
`;

exports[`themes/tokens/color/utilities/convert > oklchToRgb > black 1`] = `
{
  "b": 0,
  "g": 0,
  "mode": "rgb",
  "r": 0,
}
`;

exports[`themes/tokens/color/utilities/convert > oklchToRgb > red 1`] = `
{
  "b": -0,
  "g": 0,
  "mode": "rgb",
  "r": 1,
}
`;

exports[`themes/tokens/color/utilities/convert > oklchToRgb > white 1`] = `
{
  "b": 1,
  "g": 1,
  "mode": "rgb",
  "r": 1,
}
`;

exports[`themes/tokens/color/utilities/convert > tokenToOklchToken > hex 1`] = `
{
  "oklch": {
    "c": 0,
    "l": 1.0000000000000002,
    "mode": "oklch",
  },
  "oklchDark": undefined,
}
`;

exports[`themes/tokens/color/utilities/convert > tokenToOklchToken > hex and hexDark 1`] = `
{
  "oklch": {
    "c": 0,
    "l": 1.0000000000000002,
    "mode": "oklch",
  },
  "oklchDark": {
    "c": 0,
    "l": 0,
    "mode": "oklch",
  },
}
`;

exports[`themes/tokens/color/utilities/convert > tokenToOklchToken > oklch 1`] = `
{
  "oklch": {
    "c": 0,
    "l": 1.0000000000000002,
    "mode": "oklch",
  },
  "oklchDark": undefined,
}
`;

exports[`themes/tokens/color/utilities/convert > tokenToOklchToken > oklch and oklchDark 1`] = `
{
  "oklch": {
    "c": 0,
    "l": 1.0000000000000002,
    "mode": "oklch",
  },
  "oklchDark": {
    "c": 0,
    "l": 0,
    "mode": "oklch",
  },
}
`;
